source/opt/ir.cpp

```diff
@@ -0,0 +1,177 @@
+// Copyright (c) 2016 Google Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a
+// copy of this software and/or associated documentation files (the
+// "Materials"), to deal in the Materials without restriction, including
+// without limitation the rights to use, copy, modify, merge, publish,
+// distribute, sublicense, and/or sell copies of the Materials, and to
+// permit persons to whom the Materials are furnished to do so, subject to
+// the following conditions:
+//
+// The above copyright notice and this permission notice shall be included
+// in all copies or substantial portions of the Materials.
+//
+// MODIFICATIONS TO THIS FILE MAY MEAN IT NO LONGER ACCURATELY REFLECTS
+// KHRONOS STANDARDS. THE UNMODIFIED, NORMATIVE VERSIONS OF KHRONOS
+// SPECIFICATIONS AND HEADER INFORMATION ARE LOCATED AT
+//    https://www.khronos.org/registry/
+//
+// THE MATERIALS ARE PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
+// EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
+// MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.
+// IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY
+// CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT,
+// TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
+// MATERIALS OR THE USE OR OTHER DEALINGS IN THE MATERIALS.
+
+#include <cassert>
+
+#include "ir.h"
+#include "reflect.h"
+
+namespace spvtools {
+namespace opt {
+namespace ir {
+
+Instruction::Instruction(const spv_parsed_instruction_t& inst,
+                         std::vector<Instruction>&& dbg_line)
+    : opcode_(static_cast<SpvOp>(inst.opcode)),
+      type_id_(inst.type_id),
+      result_id_(inst.result_id),
+      dbg_line_insts_(std::move(dbg_line)) {
+  assert((!IsDebugLineInst(opcode_) || dbg_line.empty()) &&
+         "Op(No)Line attaching to Op(No)Line found");
+  for (uint32_t i = 0; i < inst.num_operands; ++i) {
+    const auto& current_payload = inst.operands[i];
+    std::vector<uint32_t> words(
+        inst.words + current_payload.offset,
+        inst.words + current_payload.offset + current_payload.num_words);
+    operands_.emplace_back(current_payload.type, std::move(words));
+  }
+}
+
+const Operand& Instruction::GetOperand(uint32_t index) const {
+  if (index >= operands_.size()) {
+    assert(0 && "operand index out of bound");
+  }
+
+  return operands_[index];
+};
+
+uint32_t Instruction::GetSingleWordOperand(uint32_t index) const {
+  if (index >= operands_.size()) {
+    // TODO(antiagainst): do better than panicking.
+    assert(0 && "operand index out of bound");
+  }
+  if (operands_[index].words.size() != 1) {
```

Comment:
The blurb for this method implies that it's usable with logical operands having more than one word.    I guess you should fix the blurb?